radix-transactions/src/manifest/static_resource_movements/types.rs

```diff
@@ -0,0 +1,545 @@
+pub enum AccountDeposit {
+    KnownFungible(ResourceAddress, FungibleBounds),
+    KnownNonFungible(ResourceAddress, NonFungibleBounds),
+    Unknown(WorktopUncertaintySource),
+}
+
+impl From<InvocationIo> for AccountDeposit {
+    fn from(value: InvocationIo) -> Self {
+        match value {
+            InvocationIo::KnownFungible(address, bound) => Self::KnownFungible(address.0, bound),
+            InvocationIo::KnownNonFungible(address, bound) => {
+                Self::KnownNonFungible(address.0, bound)
+            }
+            InvocationIo::Unknown(uncertainty) => Self::Unknown(uncertainty),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(super) enum BucketContent {
+    Fungible(FungibleResourceAddress, FungibleBounds),
+    NonFungible(NonFungibleResourceAddress, NonFungibleBounds),
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct FungibleBounds {
+    pub lower: LowerFungibleBound,
+    pub upper: UpperFungibleBound,
+}
+
+impl FungibleBounds {
+    pub fn new_exact(amount: Decimal) -> Self {
+        Self {
+            lower: LowerFungibleBound::Amount(amount),
+            upper: UpperFungibleBound::Amount(amount),
+        }
+    }
+
+    pub fn combine(&mut self, other: Self) -> Option<()> {
+        // Handling the lower bound.
+        match (self.lower, other.lower) {
+            // Two non-zero's produce a non-zero - no change needed.
+            (LowerFungibleBound::NonZero, LowerFungibleBound::NonZero) => {}
+            // Non-zero and an amount produces an amount.
+            (LowerFungibleBound::NonZero, LowerFungibleBound::Amount(amount))
+            | (LowerFungibleBound::Amount(amount), LowerFungibleBound::NonZero) => {
+                self.lower = LowerFungibleBound::Amount(amount)
+            }
+            // Two amounts get added together
+            (
+                LowerFungibleBound::Amount(ref mut self_lower_bound),
+                LowerFungibleBound::Amount(other_lower_bound),
+            ) => {
+                *self_lower_bound = self_lower_bound.checked_add(other_lower_bound)?;
+            }
+        };
+
+        // Handling the upper bound.
+        match (self.upper, other.upper) {
+            // If both upper bounds are known then the new upper bound will also be
+            // known.
+            (
+                UpperFungibleBound::Amount(ref mut self_upper_bound),
+                UpperFungibleBound::Amount(other_upper_bound),
+            ) => {
+                *self_upper_bound = self_upper_bound.checked_add(other_upper_bound)?;
+            }
+            // If either of the upper bound is unbounded then the new upper bound
+            // is also unbounded.
+            (_, UpperFungibleBound::Unbounded) | (UpperFungibleBound::Unbounded, _) => {
+                self.upper = UpperFungibleBound::Unbounded;
+            }
+        };
+
+        Some(())
+    }
+
+    pub fn increase_lower_bound(&mut self, new_lower_bound: Decimal) {
```

Comment:
(Addressed in https://github.com/radixdlt/radixdlt-scrypto/pull/1931 where we have three different semantics for handling add / take / assert against a `ResourceBound`)

Reply:
> That is the following should be an error, not a deposit of 6 XRD!

I would say that it depends on whether the worktop content is fully known at that point or not. If they're not fully known (i.e., there's a source of uncertainty), then this could totally be valid.

Reply:
> I would say that it depends on whether the worktop content is fully known at that point or not. If they're not fully known (i.e., there's a source of uncertainty), then this could totally be valid.

In my example there is no source of uncertainty - and I'm saying the output of this method is incorrect in this case.

Also, ideally any lack-of-knowledge should be handled explicitly in the model, so these assertions can interact with it appropriately.